core/common/test/format/DateTimeComponentsFormatTest.kt

```diff
@@ -268,4 +267,127 @@ class DateTimeComponentsFormatTest {
+    }
+
+    @Test
+    fun testSpecialNamedTimezones() {
+        assertTimezoneParsingBehavior(TimezoneTestData.tzPrefixes, ParseExpectation.SHOULD_PARSE_CORRECTLY)
+    }
+
+    @Test
+    fun testPrefixWithCorrectParsableOffset() {
+        val timezoneIds = generateTimezoneIds(TimezoneTestData.tzPrefixes + "", TimezoneTestData.correctParsableOffsets)
+        assertTimezoneParsingBehavior(timezoneIds, ParseExpectation.SHOULD_PARSE_CORRECTLY)
+    }
+
+    @Test
+    fun testPrefixWithIncorrectParsableOffset() {
+        val timezoneIds = generateTimezoneIds(TimezoneTestData.tzPrefixes + "", TimezoneTestData.incorrectParsableOffsets)
+        assertTimezoneParsingBehavior(timezoneIds, ParseExpectation.SHOULD_PARSE_INCORRECTLY)
+    }
+
+    @Test
+    fun testPrefixWithIncorrectUnparsableOffset() {
+        val timezoneIds = generateTimezoneIds(TimezoneTestData.tzPrefixes + "", TimezoneTestData.incorrectUnparsableOffsets)
+        assertTimezoneParsingBehavior(timezoneIds, ParseExpectation.SHOULD_FAIL_TO_PARSE)
+    }
+
+    @Test
+    fun testTimezoneDBIdentifiers() {
+        assertTimezoneParsingBehavior(TimezoneTestData.timezoneDbIdentifiers, ParseExpectation.SHOULD_PARSE_CORRECTLY)
+    }
+
+    @Test
+    fun testInvalidTimezoneIds() {
+        assertTimezoneParsingBehavior(TimezoneTestData.invalidTimezoneIds, ParseExpectation.SHOULD_FAIL_TO_PARSE)
+    }
+
+    enum class ParseExpectation {
+        SHOULD_PARSE_CORRECTLY,
+        SHOULD_PARSE_INCORRECTLY,
+        SHOULD_FAIL_TO_PARSE
+    }
+
+    private fun generateTimezoneIds(prefixes: List<String>, offsets: List<String>): List<String> = buildList {
+        for (prefix in prefixes) {
+            for (sign in listOf('+', '-')) {
+                for (offset in offsets) {
+                    add("$prefix$sign$offset")
+                }
+            }
+        }
+    }
+
+    private fun assertTimezoneParsingBehavior(zoneIds: List<String>, expectation: ParseExpectation) {
+        zoneIds.forEach { zoneId ->
+            when (expectation) {
+                ParseExpectation.SHOULD_PARSE_CORRECTLY -> {
+                    TimeZone.of(zoneId)
+                    val result = DateTimeComponents.Format { timeZoneId() }.parse(zoneId)
+                    assertEquals(zoneId, result.timeZoneId)
+                }
+
+                ParseExpectation.SHOULD_PARSE_INCORRECTLY -> {
+                    assertFailsWith<IllegalTimeZoneException> { TimeZone.of(zoneId) }
+                    val result = DateTimeComponents.Format { timeZoneId() }.parse(zoneId)
+                    assertEquals(zoneId, result.timeZoneId)
+                }
+
+                ParseExpectation.SHOULD_FAIL_TO_PARSE -> {
+                    assertFailsWith<DateTimeFormatException> {
+                        DateTimeComponents.Format { timeZoneId() }.parse(zoneId)
+                    }
+                }
+            }
+        }
+    }
```

Comment:
I don't see the benefit of this being one function instead of being three functions. In all usages, this is called with a constant value for `expectation`, and
```kotlin
assertTimezoneParsingBehavior(TimezoneTestData.invalidTimezoneIds, ParseExpectation.SHOULD_FAIL_TO_PARSE)
```
is longer than
```kotlin
assertNonParseableAsTimeZone(TimezoneTestData.invalidTimezoneIds)
```

This is not much shorter than
```kotlin
TimezoneTestData.invalidTimezoneIds.forEach(::assertNonParseableAsTimeZone)
```
which I believe to be the clearest version, given that we also pass one-element lists to `assertTimezoneParsingBehavior` in some places.